### PR TITLE
Always retry connection reset error in webhook

### PR DESCRIPTION
**Pull request by crassirostris**
Fixes https://github.com/kubernetes/kubernetes/issues/52909

Audit logging uses webhook to send events to the backend and currently even a little blip in networking can cause several hundreds of events to be lost. This PR adds an additional check, that is similar to [the one in the rest package](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request.go#L657), but ignores the fact that the request is not GET and always retries "Connection reset by peers" error.

```release-note
Webhook always retries connection reset error.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 1]
```diff
@@ -112,6 +113,11 @@ func WithExponentialBackoff(initialBackoff time.Duration, webhookFn func() error
 		if _, shouldRetry := apierrors.SuggestsClientDelay(err); shouldRetry {
 			return false, nil
 		}
+		// "Connection reset by peer" is usually a transient error, but it's not
+		// retried by the rest client because the request is not GET.
+		if net.IsConnectionReset(err) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: shouldn't this move up to the first if-clause?

**Comment by crassirostris** [comment]
I feel that this is kind of special case, requiring a separate explanation.
This is not a delay, but an outage

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: I would move IsConnectionReset and also IsInternalError up. Both are unconditional, as the others. Then for all other errors we call `SuggestsClientDelay`. Now in the current shape it looks like the order of the tests matter. But it doesn't.

**Comment by crassirostris** [comment]
@sttts PTAL

I removed the `IsServerTimeout` check, b/c it's included in the `SuggestsClientDelay` function

**Comment by sttts** [comment]
Which other webhook does this apply to? Admission? Authorization?

/cc @liggitt

**Comment by crassirostris** [comment]
> Which other webhook does this apply to? Admission? Authorization?

All of them, yes

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 2]
```diff
@@ -104,17 +105,14 @@ func WithExponentialBackoff(initialBackoff time.Duration, webhookFn func() error
 	var err error
 	wait.ExponentialBackoff(backoff, func() (bool, error) {
 		err = webhookFn()
-		// these errors indicate a need to retry an authentication check
-		if apierrors.IsServerTimeout(err) || apierrors.IsTimeout(err) || apierrors.IsTooManyRequests(err) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: why did you remove `IsServerTimeout`?

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: Because it's already checked in `SuggestsClientDelay`. I can return it, if you want the check to be more explicit

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 3]
```diff
@@ -104,17 +105,14 @@ func WithExponentialBackoff(initialBackoff time.Duration, webhookFn func() error
...
+		// these errors indicate a transient error that should be retried.
+		if net.IsConnectionReset(err) || apierrors.IsInternalError(err) || apierrors.IsTimeout(err) || apierrors.IsTooManyRequests(err) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: doesn't client-go already turn connection errors into internal server errors?

https://github.com/kubernetes/kubernetes/blob/4a6fec738b16db1f26ca968ab22984aec4e817ca/staging/src/k8s.io/client-go/rest/request.go#L652-L667

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: client-go only retries on GET calls because it doesn't want to replay non-idempotent calls. this code should not assume all webhook calls are idempotent, either. seems like it would be better to be able to indicate a request is idempotent

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: Only if it's a `GET`, but e.g. request with audit logs is `POST`

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: right... this shouldn't assume all webhook calls are idempotent

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: Writing audit logs is not an idempotent call, but it's still better to write audit logs twice than loose, esp. considering that audit logs have ids in it, that allow deduplication

> seems like it would be better to be able to indicate a request is idempotent

Is there really a need to do so? We still cannot guarantee that a request hasn't been processed if e.g. a timeout is returned. Additionally, for all known to me webhooks (auth[nz], audit) I see no benefit in not retrying connection resets

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: I mean, if we really care about a request being idempotent in this case, we shouldn't unconditionally retry e.g. 5XX. However, nobody complained about this code before, so I think it's safe to assume that retrying is more important than idempotence

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: > I mean, if we really care about a request being idempotent in this case, we shouldn't unconditionally retry e.g. 5XX. However, nobody complained about this code before, so I think it's safe to assume that retrying is more important than idempotence, WDYT?

Hmm, that's probably true

**Comment by crassirostris** [comment]
@sttts @liggitt Do you want me to change something?

**Comment by ericchiang** [comment]
At this point, would it be easier to list the cases where we don't want to retry?

**Comment by crassirostris** [comment]
@ericchiang 

> At this point, would it be easier to list the cases where we don't want to retry?

I'm not sure that is the right strategy: IMO it's better to be conservative. I.e. if some new not anticipated problem occurs, it's better to be as vocal about it as possible. For example, backend suddenly starts to reply with 418. It's better to notice this in monitoring and fix earlier, than retry each request hoping backend will stop acting as a teapot.

**Comment by sttts** [comment]
I am happy with the current shape, but lack the knowledge about the webhook context to forsee any side-effects. @liggitt any doubt this can go in?

**Comment by liggitt** [comment]
I think this is ok
